### PR TITLE
More social media share buttons

### PR DESCRIPTION
**Pull request by adam-p-king**
Social media sharing is currently limited to Twitter alone. I propose adding Facebook and Google+ to the mix with customisable fields to populate the share URL querystrings. Facebook, given the limitations of the API, is currently limited to the sharer.php method due to App Ids requiring valid domains to be provided. Twitter, however, can be upgraded to the [Web Intents framework ](https://dev.twitter.com/web/intents) and still requires no account permissions to be granted by the user in order to share.

Google, as far as I can tell, is limited to a URL to share only and thus cannot be pre-populated with content.

More information on the matter can be found in issue #38.

### CODE REVIEW

File: library.json [hunk 1]
```diff
@@ -1,9 +1,10 @@
+
 {
   "title": "Course Presentation",
   "contentType": "Library",
   "majorVersion": 1,
-  "minorVersion": 16,
-  "patchVersion": 1,
+  "minorVersion": 17,
+  "patchVersion": 3,
```
Comment (falcon-git) [inline library.json]: patchVersion probably should be 0?

Comment (falcon-git) [inline library.json]: (I'll try to get this PR reviewed tomorrow. Looks promising!)

Comment (adam-p-king) [inline library.json]: Sure, I bumped it during my commits as I patched a few fixes, thanks for taking a look!

Comment (falcon-git) [inline library.json]: If you use the development mode you don't need any extra bumping of it

**Comment by falcon-git** [comment]
Tested it now, and it works well. Especially the Facebook sharing is a very nice addition. Good job.

Details:
 - All links are empty in the authoring tool. Might confuse authors. Add a comment about them being detected automatically? Are they needed as fields at all?

 - Group all social stuff in one new group to not fill up too much in the behaviour settings group. Also there are ways now in H5P to make a checkbox show and hide other fields. The showWhen library for instance, but this is not required.

 - Icons need a text for readspeakers. aria-label for instance (Share on Twitter)

 - Google+ sharing didn't seem to work. No link?

**Comment by falcon-git** [comment]
What do you think?

**Comment by adam-p-king** [comment]
Yeah I agree with that, I've made most of the changes; just the documentation on showWhen is very basic. For instance it states the conditions in the rules object should be a string, but whether that applies to boolean values isn't stated. I tried both, although the options are always hidden, the html elements do have the showWhen classes applied to them however. It also states that the "field" proper is 

> the path and name of the field

Yet I'm unsure if that would be a fully qualified path or local in that object, e.g.
`override.social.facebookShareOverrides.showFacebookShare`
or 
`showFacebookShare`

Apologies for the delay, I'm bouncing between projects!

**Comment by falcon-git** [comment]
Hi, I've asked the ShowWhen creator to pitch in here. Hopefully he'll be able to do that tomorrow. He is not available today. I think ShowWhen is still under development but it should be safe to use it. We just have to figure out those things that you mention.

**Comment by fnoks** [comment]
@adam-p-king : I have updated the documentation here: https://github.com/h5p/h5p-editor-show-when. Let me know if I can do anything else to make it even clearer.

**Comment by adam-p-king** [comment]
Ah! `"../a/initialFullscreen"` this is what I needed, I couldn't tell what the delimiter for the path was (object or path syntax). I'll give it a go now, but that should solve my problems. Thanks for taking the time to update those!

**Comment by adam-p-king** [comment]
Hey, just finished up my other project so I'm free to focus on this. I tried the methods provided with odd results.  I'll post a snippet from what I'm testing right now, see if you can spot anything out of the ordinary. 

```json
"name": "social",
"type": "group",
"label": "Social Settings",
"importance": "low",
"description": "These options will let you override behaviour settings.",
"optional": true,
"fields": [
    {
        "name": "facebookShareOverrides",
        "importance": "low",
        "type": "group",
        "label": "Facebook share overrides",
        "fields": [
            {
                "name": "showFacebookShare",
                "type": "boolean",
                "label": "Display Facebook share icon",
                "importance": "low",
                "default": true
            },
            {
                "name": "facebookShareUrl",
                "type": "text",
                "label": "Share to Facebook link",
                "importance": "low",
                "default": "",
                "widget": "showWhen",
                "showWhen": {
                    "rules": [
                        {
                            "field": "override/social/facebookShareOverrides/showFacebookShare",
                            "equals": true
                        }
                    ]
                }
            }
        ]
    }
]
```

The odd results I speak of generally mean that items do not disappear when they meet the requirements (after save/refresh) and bundle together a lot closer than normal.

An image of the effects can be found [here](http://i.imgur.com/eOPGxIy.png)

**Comment by adam-p-king** [comment]
Hey, just finished up my other project so I'm free to focus on this. I tried the methods provided with odd 
results.  I'll post a snippet from what I'm testing right now, see if you can spot anything out of the ordinary. 

It did add the showWhen class to the elements specified but the effect was wrong.

```json
"name": "social",
"type": "group",
"label": "Social Settings",
"importance": "low",
"description": "These options will let you override social media sharing settings.",
"optional": true,
"fields": [
    {
        "name": "facebookShareOverrides",
        "importance": "low",
        "type": "group",
        "label": "Facebook share overrides",
        "fields": [
            {
                "name": "showFacebookShare",
                "type": "boolean",
                "label": "Display Facebook share icon",
                "importance": "low",
                "default": true
            },
            {
                "name": "facebookShareUrl",
                "type": "text",
                "label": "Share to Facebook link",
                "importance": "low",
                "default": "",
                "widget": "showWhen",
                "showWhen": {
                    "rules": [
                        {
                            "field": "override/social/facebookShareOverrides/showFacebookShare",
                            "equals": true
                        }
                    ]
                }
            }
        ]
    }
]
```

The odd results I speak of generally mean that items do not disappear when they meet the requirements (after save/refresh) and bundle together a lot closer than normal.

An image of the effects can be found [here](http://i.imgur.com/eOPGxIy.png)

I've tried using the detach method but that lead to everything being hidden, regardless of the state the rule target entity was in; I do highly believe this is my doing though.

**Comment by falcon-git** [comment]
Read through the code now and it seems to have fixed many things. Not sure about "All links are empty in the authoring tool. Might confuse authors. Add a comment about them being detected automatically? Are they needed as fields at all?"

There are some @domain and other placeholders there, were it there all the time, maybe a description saying that they will be replaced with a url to the site? By the way, I think perhaps there might be a variable available for site url in the h5pintegration object, but I'm not sure.

**Comment by falcon-git** [comment]
I've pinged fnoks to get some input from him on show when. The github notification from this issue ended up in a bad place in our support system.

**Comment by fnoks** [comment]
Here is my suggestion on how these fields should be organized:
```
{
  "name": "social",
  "type": "group",
  "label": "Social Settings",
  "importance": "low",
  "description": "These options will let you override social behaviour settings. Empty values will be filled in automatically if a link is provided, otherwise all values will be generated.",
  "optional": true,
  "fields": [
    {
      "name": "showFacebookShare",
      "type": "boolean",
      "label": "Display Facebook share icon",
      "importance": "low",
      "default": true
    },
    {
      "name": "facebookShareOverrides",
      "importance": "low",
      "type": "group",
      "label": "Facebook share settings",
      "widget": "showWhen",
      "showWhen": {
        "rules": [
          {
            "field": "showFacebookShare",
            "equals": true
          }
        ]
      },
      "fields": [
        {
          "name": "facebookShareUrl",
          "type": "text",
          "label": "Share to Facebook link",
          "importance": "low",
          "default": "@domain"
        },
        {
          "name": "facebookShareTitle",
          "type": "text",
          "label": "Share to Facebook title",
          "importance": "low",
          "default": "@domain"
        },
        {
          "name": "facebookShareQuote",
          "type": "text",
          "label": "Share to Facebook quote",
          "importance": "low",
          "default": "I scored @percentage on a task at @url."
        },
        {
          "name": "facebookShareDescription",
          "type": "text",
          "label": "Share to Facebook description",
          "importance": "low",
          "default": ""
        }
      ]
    },
    {
      "name": "showTwitterShare",
      "type": "boolean",
      "label": "Display Twitter share icon",
      "importance": "low",
      "default": false
    },
    {
      "name": "twitterShareOverrides",
      "importance": "low",
      "type": "group",
      "label": "Twitter share settings",
      "widget": "showWhen",
      "showWhen": {
        "rules": [
          {
            "field": "showTwitterShare",
            "equals": true
          }
        ]
      },
      "fields": [
        {
          "name": "twitterShareStatement",
          "type": "text",
          "label": "Share to Twitter statement",
          "importance": "low",
          "default": "I scored @percentage on a task at @domain."
        },
        {
          "name": "twitterShareUrl",
          "type": "text",
          "label": "Share to Twitter link",
          "importance": "low",
          "default": "@domain"
        },
        {
          "name": "twitterShareHashtags",
          "type": "text",
          "label": "Share to Twitter hashtags",
          "importance": "low",
          "default": "h5p, course"
        }
      ]
    },
    {
      "name": "showGoogleShare",
      "type": "boolean",
      "label": "Display Google+ share icon",
      "importance": "low",
      "default": false
    },
    {
      "name": "googleShareOverrides",
      "importance": "low",
      "type": "group",
      "label": "Google+ share settings",
      "widget": "showWhen",
      "showWhen": {
        "rules": [
          {
            "field": "showGoogleShare",
            "equals": true
          }
        ]
      },
      "fields": [
        {
          "name": "googleShareUrl",
          "type": "text",
          "label": "Share to Google link",
          "importance": "low",
          "default": "@domain"
        }
      ]
    }
  ]
}
```

i.e:
- I have placed the boolean's outside the group, and show the group only if the checkbox is disabled.
- I also changed some of the texts (using "settings" insterad of "overrides")

**Comment by adam-p-king** [comment]
That's alright, I didn't get notified at all so I'm a bit late getting back to this, also! I'll go about implementing the change today as suggested and see where it takes us, thanks for the input.

**Comment by adam-p-king** [comment]
Finally found time to get back to this lately and I have to say, the suggested semantics worked without a hitch with minor modifications to the code afterwards; thanks for the help there @fnoks.

**Comment by adam-p-king** [comment]
Of course, the template literals will be removed. By the looks of it, Facebook have removed the Title, Description and Image parameters from the sharer.php file I'm using to share and instead use the OG tags on the page provided via URL so I will definitely remove those options too.

As for the second point, though, I think it's best if we leave that as default but give people a choice; a common scenario I can think of is if the URL is shared from a page the public may not have access to, for example a Moodle course or content that is walled-off behind a user login system.

**Comment by fnoks** [comment]
If we leave the URL configurable, the default value should be the current page's URL, and could be set using e.g @currentPageUrl as default value (which should be replaced with window.location.href in code). I don't think we need additional replacement variables (i.e remove support for @domain and @url)

**Comment by adam-p-king** [comment]
The template literals and redundant Facebook sharer properties are gone and both **@domain** and **@url** no longer exist, the defaults are now **@currentpageurl** as suggested

**Pull request merged**